include/deal.II/lac/scalapack.h

```diff
@@ -0,0 +1,478 @@
+// ---------------------------------------------------------------------
+//
+// Copyright (C) 2017 by the deal.II authors
+//
+// This file is part of the deal.II library.
+//
+// The deal.II library is free software; you can use it, redistribute
+// it, and/or modify it under the terms of the GNU Lesser General
+// Public License as published by the Free Software Foundation; either
+// version 2.1 of the License, or (at your option) any later version.
+// The full text of the license can be found in the file LICENSE at
+// the top level of the deal.II distribution.
+//
+// ---------------------------------------------------------------------
+
+#ifndef dealii_scalapack_h
+#define dealii_scalapack_h
+
+#include <deal.II/base/config.h>
+
+#ifdef DEAL_II_WITH_SCALAPACK
+
+#include <deal.II/base/exceptions.h>
+#include <deal.II/lac/full_matrix.h>
+#include <deal.II/lac/lapack_support.h>
+#include <deal.II/base/mpi.h>
+#include <mpi.h>
+
+#include <memory>
+
+DEAL_II_NAMESPACE_OPEN
+
+
+/**
+ * Forward declaration of class ScaLAPACKMatrix
+ */
+template <typename NumberType> class ScaLAPACKMatrix;
+
+
+/**
+ * A class taking care of setting up the process grid for BLACS
+ * ScaLapack matrices will have shared pointer of this class to perform block-cyclic distribution
+ *
+ * @author Benjamin Brands, 2017
+ */
+class ProcessGrid
+{
+public:
+
+  /**
+   * Declare class ScaLAPACK as friend to provide access to private members, e.g. the MPI Communicator
+   */
+  template <typename NumberType> friend class ScaLAPACKMatrix;
+
+  /*
+   * Constructor for a process grid for a given @p mpi_communicator
+   * The pair @p grid_dimensions contains the user-defined numbers of process rows and columns
```

Comment:
does this mean the product of the two dimensions needs to be equal to the number of ranks? If yes, please state this.

Reply:
no, it does not. It should be obviously less or equal to, i will update. 

p.s. we try to keep it working fine even when number of cores is slightly more, thus distinction between active and in-active cores.